zerver/lib/streams.py

```diff
@@ -757,27 +757,57 @@ def can_access_stream_history_by_id(user_profile: UserProfile, stream_id: int) -
     return can_access_stream_history(user_profile, stream)
 
 
-def can_remove_subscribers_from_stream(
-    stream: Stream, user_profile: UserProfile, sub: Subscription | None
+def bulk_can_remove_subscribers_from_streams(
+    streams: list[Stream], user_profile: UserProfile
 ) -> bool:
-    if not check_basic_stream_access(user_profile, stream, sub, allow_realm_admin=True):
-        return False
-
     # Optimization for the organization administrator code path. We
     # don't explicitly grant realm admins this permission, but admins
     # implicitly have the can_administer_channel_group permission for
-    # all accessible channels.
+    # all channels. For channels the user cannot access, they can
+    # unsubscribe other users if they have permission to administer
+    # that channel.
     if user_profile.is_realm_admin:
         return True
 
-    group_allowed_to_remove_subscribers = stream.can_remove_subscribers_group
-    assert group_allowed_to_remove_subscribers is not None
-    return user_has_permission_for_group_setting(
-        group_allowed_to_remove_subscribers,
-        user_profile,
-        Stream.stream_permission_group_settings["can_remove_subscribers_group"],
+    user_recursive_group_ids = set(
+        get_recursive_membership_groups(user_profile).values_list("id", flat=True)
     )
 
+    # We check this before basic access since for the channels the user
+    # cannot access, they can unsubscribe other users if they have
+    # permission to administer that channel.
+    is_allowed_to_administer_all_streams = True
+    for stream in streams:
+        group_allowed_to_administer_channel_id = stream.can_administer_channel_group_id
+        assert group_allowed_to_administer_channel_id is not None
+        if group_allowed_to_administer_channel_id not in user_recursive_group_ids:
+            is_allowed_to_administer_all_streams = False
+            break
+
+    if is_allowed_to_administer_all_streams:
+        return True
+
+    existing_recipient_ids = [stream.recipient_id for stream in streams]
+    subs = Subscription.objects.filter(
+        user_profile=user_profile, recipient_id__in=existing_recipient_ids, active=True
+    )
+    sub_map = {sub.recipient_id: sub for sub in subs}
```

Comment:
This can be optimized to not fetch the entire `Subscription` objects, which aren't tiny. `check_basic_stream_access` only needs to know if the user is subscribed.

I think it can be refactored to accept an `is_subscribed` boolean (existing callers can pass `sub is not None`), and then we can just fetch the `.values_list("recipient_id", flat=True)`.

Reply:
[Fixed](https://github.com/zulip/zulip/compare/bedb600897f6e8efb7039117b7d3962128973f90..59947764093223865356a5c6fda95cafb885568f)